File: src/js/core/stacktrace.js
/*
 * stacktrace.js
 *
 * Copyright (C) 2008 - 2011 by Joshua S. English. This document is the
 * intellectual property of Joshua S. English, all rights reserved.
 *
 * The Asgard Native API core library to deal with the JavaScript stack.
 *
 * Written by Josh English.
 */

// define stacktrace functions

function stacktrace_getStacktrace(error)
{
	var callstack = null;
	var currentFunction = '';
	var currentFunctionName = '';
	var currentFunctionString = '';
 
	try {
		if(error.stack) { // firefox
			callstack = error.stack.split('\n');
		}
		else {
			callstack = new Array();

			currentFunction = arguments.callee.caller;

			while(currentFunction) {
				currentFunctionString = currentFunction.toString();

				currentFunctionName = currentFunctionString.substring(
						currentFunctionString.indexOf(
							"function") + 8,
						currentFunctionString.indexOf(''));

				if(currentFunctionName = undefined) {
					currentFunctionName = 'unknown';
				}

				callstack.push(currentFunctionName);

				currentFunction = currentFunction.caller;
			}
		}
	}
	catch(localError) {
		print('error - stacktrace_getStacktrace() :: ' + localError + ' (' +
					localError.description + ')');
	}

	return callstack;
}

function stacktrace_printStacktrace(error)
{
	var ii = 0;
	var trace = null;

	try {
		trace = stacktrace_getStacktrace(error);

		if((trace != undefined) && (trace != null)) {
			for(ii = 0; ii < trace.length; ii++) {
				print('[stack ' + ii + '] ' + trace[ii]);
			}
		}
	}
	catch(localError) {
		print('error - stacktrace_printStacktrace() :: ' + localError + ' (' +
					localError.description + ')');
	}
}

File: src/c/core/system/stacktrace.c
/*
 * stacktrace.c
 *
 * Copyright (C) 2008 - 2011 by Joshua S. English. This document is the
 * intellectual property of Joshua S. English, all rights reserved.
 *
 * The Asgard system-stacktrace function library.
 *
 * Written by Josh English.
 */

// preprocessor directives

#include "core/core.h"
#define _CORE_SYSTEM_COMPONENT
#include "core/system/stacktrace.h"
#define _CORE_SYSTEM_PLUGIN_COMPONENT
#include "core/system/plugin/stacktrace_android.h"
#include "core/system/plugin/stacktrace_linux.h"
#include "core/system/plugin/stacktrace_windows.h"

// define stacktrace public functions

int stacktrace_getStack(char **stackString, int *stackStringLength)
{
#if (defined(__linux__) || defined(__APPLE__)) && !defined(__ANDROID__)
	return stacktrace_getStackLinux(stackString, stackStringLength);
#elif defined(WIN32)
	return stacktrace_getStackWindows(stackString, stackStringLength);
#elif defined(__ANDROID__)
	return stacktrace_getStackAndroid(stackString, stackStringLength);
#else // - no plugin available -
	return -1;
#endif // - plugins -
}

